### PR TITLE
Adding thread protection for the local data structures

### PR DESCRIPTION
**Pull request by tanmayv25**

### CODE REVIEW

File: src/clients/python/__init__.py [hunk 1]
```diff
@@ -1501,6 +1504,7 @@ def async_run_with_cb(self, callback, inputs, outputs, batch_size=1, flags=0):
         wrapped_cb = partial(self._async_callback_wrapper, self._callback_resources_dict_id, callback)
         c_cb = _async_run_callback_prototype(wrapped_cb)
 
+        self._lock.acquire()
         # Run asynchronous inference...
```
Comment (deadeyegoodwin) [inline src/clients/python/__init__.py]: Use a try .. finally to make sure you always release the lock. Here is an exception is raised below the lock won't be released.

File: src/clients/python/__init__.py [hunk 2]
```diff
@@ -1548,10 +1553,11 @@ def get_async_run_results(self, request_id, wait):
         err = c_void_p(_crequest_infer_ctx_get_async_run_results(
             self._ctx, byref(c_is_ready), request_id, wait))
 
-
+        self._lock.acquire()
```
Comment (deadeyegoodwin) [inline src/clients/python/__init__.py]: Use try...finally.

Why does this acquire need to be here. Can't it be moved down to just before "requested_outputs = ... "

File: src/clients/python/__init__.py (hunk 1 above)
Comment (tanmayv25) [reply src/clients/python/__init__.py]: using "with self._lock:" block to ensure the lock is released in all cases.

File: src/clients/python/__init__.py (hunk 2 above)
Comment (tanmayv25) [reply src/clients/python/__init__.py]: Yes, we can allow threads to race for updating _last_request_id. Fixed.

**Comment by deadeyegoodwin** [comment]
@tanmayv25 you need to rerun CI.

**Pull request merged**